lib/auth/sessions.go

```diff
@@ -544,6 +539,55 @@ func (a *Server) CreateAppSessionFromReq(ctx context.Context, req NewAppSessionR
 		return nil, trace.Wrap(err)
 	}
 
+	// Check if any role requires device trust
+	deviceTrustRequired := false
+	for _, role := range checker.Roles() {
+		mode := role.GetOptions().DeviceTrustMode
+
+		if mode == constants.DeviceTrustModeRequired ||
+			mode == constants.DeviceTrustModeRequiredForHumans {
+			deviceTrustRequired = true
+			break
+		}
+	}
+
+	// Validate that the request has the deviceID extension; non-empty value means user authenticated with that device with CA
+	if deviceTrustRequired {
+		if req.DeviceExtensions.DeviceID == "" {
+			errMsg := fmt.Sprintf("access to application %s requires a trusted device", req.AppName)
+			a.emitter.EmitAuditEvent(ctx, &apievents.AuthAttempt{
+				Metadata: apievents.Metadata{
+					Type:        events.AuthAttemptEvent,
+					Code:        events.AuthAttemptFailureCode,
+					ClusterName: req.ClusterName,
+				},
+				ServerMetadata: apievents.ServerMetadata{
+					ServerVersion:   teleport.Version,
+					ServerID:        a.ServerID,
+					ServerNamespace: apidefaults.Namespace,
+				},
```

Comment:
Should we omit `ServerMetadata` here since we know the access attempt was for an app, and we have app metadata below?